Покоритель баз данных

Dimaggio-IT opened the pull request:
---
:mortar_board: [Покоритель баз данных](https://up.htmlacademy.ru/nodejs-api/6/user/2256781/tasks/9)

Dimaggio-IT committed:
Добавил первую модель User. WIP

 Интерфейсы документа наследуется от `Document` (`mongoose.Document`) и
интерфейса/типа сущности. В этом шаге мы определяем интерфейс `UserDocument`, который наследуется от типа `User` и `mongoose.Document`.

После подготовки интерфейса займёмся описанием схемы. Для создания схемы применяется конструктор `Schema`. В схему перечислим все поля и их типы.

Последним шагом создадим модель. Для создания модели следует передать в метод `model` подготовленный интерфейс (в параметр типа) и саму схему. Всё, этого достаточно для записи и извлечения данных из базы.

В качестве теста в методе `init` класса `Application` создал новую запись в MongoDB:

```
 const user = await UserModel.create({
   email: 'test@email.local',
   avatarPath: 'keks.jpg',
   firstname: 'Keks',
   lastname: 'Unknown'
  });
```

и выполнить модуль.

Dimaggio-IT committed:
Расширил интерфейс UserDocument. WIP

чтобы заполнялись поля `createdAt` и `updatedAt` типа `Date` которые
были добавлены в интерфейс IUserDocument, нужно при создании
схемы указать ключ timestamps` со значением `true`

Dimaggio-IT committed:
Добавил уникальность для поля email. WIP

использована расширенная запись при определении поля на примере  `email`
`
email: {
    type: String,
    unique: true,
  }
`

Dimaggio-IT committed:
Задействовал встроенную валидацию. WIP

используются валидаторы mongoose
- match
- minLength

чтобы проверить работу валидатора можно заюзать методы `validate` или `validateSync`

```
const user = new UserModel({
  email: 'test@emailru',
  avatarPath: 'keks.jpg',
  firstname: '2',
  lastname: 'Unknown'
});

 const error = user.validateSync();
 console.log(error);
```

после того как код выполнился ловим ошибку:
Error: User validation failed: email: Email is incorrect, firstname: Min length for firstname is 2

Dimaggio-IT committed:
Установил typegoose и добавил сущность User. WIP

основная идея его в том, что вместо того, чтобы делать схему и затем делать на её основе (+ещё интерфейс желательно задать) модель мы можем определять модели и схемы с помощью классов. При этом класс будет всего лишь один. В нём опишем модель, а всю необходимую информацию для схемы извлечём за счёт декораторов, которые предоставляет пакет `@typegoose/typegoose`

после добавления сущности UserEntity схема и интерфейс `UserDocument`
больше не потребуются

Dimaggio-IT committed:
Добавил определение модели и схемы. WIP

декоратором `@prop`следует пометить свойства класса, значения которых должны сохраняться в базе данных

@prop`описываем схему как бы
поверх класса. У класса могут быть и другие свойства, которые не должны
сохраняться в базу данных (их не нужно помечать @prop)

`@prop` в качестве аргумента принимает объект настроек (см. код в коммите)

чтобы создать на основании класса модель в typegoose` предусмотрена
специальная функция — `getModelForClass`

ElkostVN commented on src/shared/libs/database-client/mongo.database-client.ts:
```diff
@@ -0,0 +1,58 @@
+
+@injectable()
+export class MongoDatabaseClient implements IDatabaseClient {
+  private isConnected: boolean;
```
private isConnected = false;

ElkostVN commented on src/shared/libs/database-client/mongo.database-client.ts:
```diff
@@ -0,0 +1,58 @@
+    this.isConnected = false;
+  }
+
+  public isConnectedToDatabase() {
```
public get isConnectedToDatabase() {
  return this.isConnected;
}

Dimaggio-IT replied:
ок

ElkostVN commented on src/shared/libs/database-client/mongo.database-client.ts:
```diff
@@ -0,0 +1,58 @@
+      throw new Error('Not connected to the database');
+    }
+
+    await this.mongoose.disconnect?.();
```
if (!(this.mongoose instanceof Mongoose)) { throw new Error('Instance not defined') }

Dimaggio-IT replied:
ок

Dimaggio-IT committed:
исправления по комментариям наставника

Dimaggio-IT committed:
Добавил поля createdAt, updatedAt. WIP

унаследовался от класса`TimeStamps` — это абстрактный класс, который определяет два нужных нам поля createdAt, updatedAt

Dimaggio-IT committed:
Добавил поля _id и id. WIP

поля _id и id потребуются в каждой модели, чтобы каждый раз их не добавлять
в модели или не добавлять эти поля в классы(здесь на примере UserEntity) можно
сделать "финт ушами", а именно: смержить интерфейс и класс. Т.е. если
объявить интерфейс и класс с одинаковым именем в TypeScript, будет выполнена операция объединения типов (type merging). В результате, и интерфейс, и класс будут содержать объединенные свойства и методы и самое вкусное - их не нужно явно нигде прописывать )

Dimaggio-IT committed:
Добавил методы в UserEntity. WIP

Пароль для пользователя будет устанавливаться с помощью метода `setPassword` - параметры у метода пароль и соль

добавил функцию хелпер `createSHA256`, где юзается`createHmac` - ему
нужно передать алгоритм хеширования и соль (используется алгоритм `sha256` - т.к. он быстрый и надёжный)

Dimaggio-IT committed:
Добавил класс CreateUserDTO

DTO (Data Transfer Object или по русски Oбъект переноса данных) - ещё один паттерн проектирования.

Его применяют для описания контрактов передачи данных между слоями приложения или подсистемами приложения.

Чтобы создать в базе данных новый документ с информацией о пользователе нужно передать определённы набор данных. Мы с ним разбирались при описании `UserEntity`. Этот набор может приходить из разных мест: от
клиента (с фронтенда), из другого слоя приложения и так далее. DTO позволяет добавить согласованность. Мы определяем «контракт» и поддерживаем его на разных слоях приложения.

DTO опишем в виде класса. Это чаще удобней, чем просто интерфейс/или тип. В нём
перечислим все поля, которые необходимы для создания нового документа пользователя.

DTO всегда будет несколько. Поэтому удобней сразу завести для них отдельную директорию внутри определённого модуля.

ПРИМЕЧАНИЕ: В 1С-ке есть подобный механизм правда с уклоном в сторону XML - называется XDTO. XDTO - это механизм, разработанный фирмой “1С” для обмена данными с другими программными системами посредством XML, позволяющий на уровне языка 1С оперировать не узлами XML, а прикладными понятиями “Сотрудник”, “Счет” и привычными встроенными типами (“ТаблицаЗначений”, “СправочникСсылка” и т.п.).

Dimaggio-IT committed:
Добавил интерфейс UserService

каждый модуль (например, `User`) будет обладать собственным сервисом для взаимодействия с базой данных. В сервисе `User` будут ф-ии
- create
- findByEmail
- findOrCreate

Dimaggio-IT committed:
Добавил DefaultUserService. WIP

реализован метод `create`. Он отвечает за создание нового документа в базе данных.

Для создания документа в БД потребуется модель. Эту модель создана с
помощью класса `getModelForClass`

Создадим сущность `UserEntity`. Мы можем этого не делать, а сразу передать все данные в метод `create` модели, но у нас здесь

интересный кейс с паролем. Для создания пароля мы предусмотрели отдельный метод `setPassword`. Соответственно, чтобы им воспользоваться,
нам потребуется создать экземпляр `UserEntity`.

WIP: Имплементирован только один метод интерфейса
`UserService`. Линтер ругается.

Dimaggio-IT committed:
Разделил общий контейнер и добавил внедрение зависимостей в DefaultUserService. WIP

использовал у библиотеки `inversify` есть метод `merge()`. Метод `merge()` принимает два и более контейнеров зависимостей и собирает их в один общий.

WIP: Имплементирован только один метод интерфейса `UserService`.
Линтер ругается.

Dimaggio-IT committed:
Добавил внедрение модели в зависимости. WIP

Класс `DefaultUserService` напрямую обращался к модели `UserModel`.
Понизил эту связь и внедрил модель в качестве зависимости

WIP: Имплементирован только один метод интерфейса `UserService`.
Линтер ругается.

Dimaggio-IT committed:
Добавил имплементацию методов IUserService

имплементировал методы  в классе `DefaultUserService`
- findByEmail
- findOrCreate

Метод `findOrCreate` совмещает работу двух методов. Сначала в нём
выполняется поиск по `email`, а затем проверка результата. Если
документ удалось найти, он станет результатом выполнения метода
`findOrCreate`. Если нет, эстафетная палочка передаётся методу
`create`

Dimaggio-IT committed:
Добавил сущность OfferEntity

по аналогии с UserEntity

однако, `OfferEntity` необычна тем, что имеет связь с другим документом, а именно `UserEntity`

чуть порефачил
- `UserEntity`
- разнёс типы чуть по другим файлам

Dimaggio-IT committed:
Добавил простой консоль логгер

Dimaggio-IT committed:
Обновил команду import

Основная цель этого коммита, данные из сгенерированного файла на
основе данных мокового json-сервера (файл:  test-data.tsv) записать в
базу данных. Записываться данные из файла должны в 2 таблицы: users
и offers. Каждая запись из таблицы offers имеет ссылку на одну запись
из таблцицы users, следовательно сначала записывается сущность `user`,
а затем ссылка этой сущности используется в сущности `offer` в поле
`userId`

Данные для подключения к базе попросим напрямую у пользователя, однако
если ничего не ввёл возьму из константы DEFAULT_COMMAND_SETTINGS

Dimaggio-IT committed:
исправил ошибки при валидации, которые возникали в базе данных
при проверке данных при записи

Dimaggio-IT committed:
удалил вопросы из кода
